File: antia_numerical_methods/f77/xmp4/median.f
!       To calculate median, mean and higher moments of a distribution
      PROGRAM MEDIAN
!      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION X(500)
      DATA (X(I),I=1,21)/49,73,46,45,67,47,28,49,61,71,49,57,52,79,
     1    40,41,55,69,41,71,40/

51    FORMAT(' Median =',F8.2,',  Mean =',F8.2,',  Standard deviation ='
     1       ,F8.2,/,'  Skewness =',F8.2,',  Kurtosis =',F8.2)

      N=21
      CALL SHSORT(X,N)
      AMED=X((N+1)/2)
      AMEAN=0.0
      DO I=1,N
        AMEAN=AMEAN+X(I)
      ENDDO
      AMEAN=AMEAN/N
      SIG=0.0
      DO I=1,N
        SIG=SIG+(X(I)-AMEAN)**2
      ENDDO
      SIG=SQRT(SIG/(N-1))
      SKEW=0.0
      AKER=0.0
      DO I=1,N
        SKEW=SKEW+((X(I)-AMEAN)/SIG)**3
        AKER=AKER+((X(I)-AMEAN)/SIG)**4
      ENDDO
      SKEW=SKEW/(N-1)
      AKER=AKER/(N-1)-3
      WRITE(6,51) AMED,AMEAN,SIG,SKEW,AKER
      END

!       -------------------------------------------------------

!	To sort an array of length N in ascending order, using shell sort algorithm
!
!	A : (input/output) Real array of length N, which is to be sorted
!               The sorted array will be overwritten on A
!	N : (input) No. of elements of array A to be sorted
!
!	Required routines : None

      SUBROUTINE SHSORT(A,N)
!      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION A(N)

      L=LOG(FLOAT(N))/LOG(2.D0)+1
      M=2**L
      IF(M.GE.2*N) M=M/2
      DO J=1,L
        M=M/2
        IF(M.EQ.0) RETURN
        DO I=M+1,N
          T=A(I)
          DO I1=I,M+1,-M
            IF(A(I1-M).GT.T) THEN
              A(I1)=A(I1-M)
            ELSE
              EXIT
            ENDIF
          ENDDO
          IF(I1.LE.0) I1=I1+M
          A(I1)=T
        ENDDO
      ENDDO
      END

File: antia_numerical_methods/c/xmp/median.c
/*	To calculate median, mean and higher moments of a distribution */

#include <stdio.h>
#include <math.h>

void shsort(double a[], int n);

main()
{
	int i,n;
	double amed,amean,sig,skew,aker,r,r2;
	double x[21]={49,73,46,45,67,47,28,49,61,71,49,57,52,79,40,41,55,69,41,71,40};

	n=21; shsort(x,n);
	amed=x[(n-1)/2];
	amean=0.0;
	for(i=0; i<n; ++i) amean=amean+x[i];
	amean=amean/n;
	sig=0.0;
	for(i=0; i<n; ++i) {
		r=(x[i]-amean); sig=sig+r*r;
	}
	sig=sqrt(sig/(n-1));
	skew=0.0; aker=0.0;
	for(i=0; i<n; ++i) {
		r=(x[i]-amean)/sig; r2=r*r;
		skew=skew+r2*r; aker=aker+r2*r2;
	}
	skew=skew/(n-1); aker=aker/(n-1)-3;
	printf(" Median = %f  Mean = %f  Standard deviation = %f \n",amed,amean,sig);
	printf(" Skewness = %f  Kurtosis = %f \n",skew,aker);
}

/*       To sort an array of length n in ascending order, using shell sort algorithm

       a : (input) Real array of length n, which is to be sorted
       n : (input) No. of elements of array a to be sorted

       Required routines : None
*/

#include <math.h>

void shsort(double a[], int n)
{
	int i,j,i1,l,m;
	double an,al,t;

	an=n; l=log(an)/log(2.0)+1; al=l;
	m=pow(2.0,al);
	if(m>=2*n) m=m/2;
	while(m>0)
	{
		m=m/2;
		for(i=m; i<n; i++)
		{
			t=a[i]; i1=i;
			while(a[i1-m]>t && i1>=m)
			{
				a[i1]=a[i1-m]; i1=i1-m;
			}
			a[i1]=t;
		}
	}
}

